File: IndieLibX/scripting/SimpleTexture.py
#
# Copyright (c) 2015 Tommi Roenty   http://www.tommironty.fi/
# Licensed under The GNU Lesser General Public License, version 2.1:
#     http://opensource.org/licenses/LGPL-2.1
#
import IndieLib

class AppPy:
	def __init__(self):
		
		self.shdTxt='''[Vertex shader]
struct VsIn {
	float2 position: POSITION;
	float2 texCoord: TEXCOORD0;
};
struct VsOut {
	float4 position: POSITION;
	float2 texCoord: TEXCOORD0;
};
ROW_MAJOR float4x4 worldViewProj MVPSEMANTIC;
VsOut main(VsIn In){
	VsOut Out=(VsOut)0;
	Out.position = mul(worldViewProj, float4(In.position.x, In.position.y, 0.0, 1.0));
	Out.texCoord = In.texCoord;
	return Out;
}
[Fragment shader]
struct VsOut {
	float4 position: POSITION;
	float2 texCoord: TEXCOORD0;
};
sampler2D Base;
float4 main(VsOut IN): COLOR {
	return tex2D(Base, IN.texCoord);
}
		'''
		self.shd = IndieLib.IRenderer.GetRendererInstance().addHLSLShader(self.shdTxt, "main", "main")
		a = IndieLib.new_intArray(8)
		IndieLib.intArray_setitem(a, 0, 0)
		IndieLib.intArray_setitem(a, 1, IndieLib.TYPE_VERTEX)
		IndieLib.intArray_setitem(a, 2, IndieLib.FORMAT_FLOAT)
		IndieLib.intArray_setitem(a, 3, 2)
		IndieLib.intArray_setitem(a, 4, 0)
		IndieLib.intArray_setitem(a, 5, IndieLib.TYPE_TEXCOORD)
		IndieLib.intArray_setitem(a, 6, IndieLib.FORMAT_FLOAT)
		IndieLib.intArray_setitem(a, 7, 2)
		self.vf = IndieLib.IRenderer.GetRendererInstance().addVertexFormatV(a, 2, self.shd)
		IndieLib.delete_intArray(a)
		ss=IndieLib.IRenderer.GetRendererInstance().Getlinear()
		self.texture=IndieLib.IRenderer.GetRendererInstance().addImageLibTexture("/test.bmp", False, ss)
		self.v = IndieLib.new_floatArray(16)
		IndieLib.floatArray_setitem(self.v, 0,  1.0)
		IndieLib.floatArray_setitem(self.v, 1, -1.0)
		IndieLib.floatArray_setitem(self.v, 2,  1.0)
		IndieLib.floatArray_setitem(self.v, 3,  1.0)
		IndieLib.floatArray_setitem(self.v, 4,  1.0)
		IndieLib.floatArray_setitem(self.v, 5,  1.0)
		IndieLib.floatArray_setitem(self.v, 6,  1.0)
		IndieLib.floatArray_setitem(self.v, 7,  0.0)
		IndieLib.floatArray_setitem(self.v, 8, -1.0)
		IndieLib.floatArray_setitem(self.v, 9, -1.0)
		IndieLib.floatArray_setitem(self.v, 10,  0.0)
		IndieLib.floatArray_setitem(self.v, 11,  1.0)

		IndieLib.floatArray_setitem(self.v, 12, -1.0)
		IndieLib.floatArray_setitem(self.v, 13,  1.0)
		IndieLib.floatArray_setitem(self.v, 14,  0.0)
		IndieLib.floatArray_setitem(self.v, 15,  0.0)
	def clear(self):
		IndieLib.delete_floatArray(self.v)

	def render(self):
		f=128.0/256.0
		IndieLib.IRenderer.GetRendererInstance().Clear(True, False, IndieLib.D3DXFROMWINEVECTOR4 (f, f, f, 1.0))
		IndieLib.IRenderer.GetRendererInstance().BeginScene()
		IndieLib.IRenderer.GetRendererInstance().setShader(self.shd)
		IndieLib.IRenderer.GetRendererInstance().setVertexFormat(self.vf)
		IndieLib.IRenderer.GetRendererInstance().setTexture("Base", self.texture)
		I=IndieLib.D3DXFROMWINEMATRIX()
		IndieLib.D3DXFROMWINEMatrixIdentity(I)
		IndieLib.IRenderer.GetRendererInstance().setShaderConstant4x4f("worldViewProj", I)
		IndieLib.IRenderer.GetRendererInstance().DrawPrimitiveUP(IndieLib.PRIM_TRIANGLE_STRIP, 2, self.v, self.v, 16)		
		IndieLib.IRenderer.GetRendererInstance().EndScene()
		IndieLib.IRenderer.GetRendererInstance().Present( )

if __name__=="__main__":
	app=AppPy()
	IndieLib.IRenderer.GetRendererInstance("SimpleTexture")		
	while ((not IndieLib.STX_Service.GetInputInstance().OnKeyPress (IndieLib.STX_KEY_ESCAPE)) and (not IndieLib.STX_Service.GetInputInstance().Quit())):
		IndieLib.STX_Service.GetInputInstance().Update()	
		app.render()
	app.clear()

File: IndieLibX/Samples/SimpleTexture/SimpleTexture.cpp
/*
  Copyright (c) 2015 Tommi Roenty   http://www.tommironty.fi/
  Licensed under The GNU Lesser General Public License, version 2.1:
      http://opensource.org/licenses/LGPL-2.1
*/
#include <Framework3/IRenderer.h>

//define STX_PRINT(...) printf(__VA_ARGS__)
//define STX_FNLN printf("%s:%s:%d\n", __FILE__,__FUNCTION__, __LINE__)

ShaderID shd=-1;
TextureID texture=-1;
VertexFormatID vf=-1;

int init(const char* aTitle)
{
	STX_FNLN;
	shd = IRenderer::GetRendererInstance()->addShaderFromFile("/SimpleTexture/SimpleTexture.shd", "mainVS", "mainPS");
	STX_FNLN;
	FormatDesc format[] =
	{
		0, TYPE_VERTEX,   FORMAT_FLOAT, 2,
		0, TYPE_TEXCOORD, FORMAT_FLOAT, 2
	};
	STX_FNLN;
	vf = IRenderer::GetRendererInstance()->addVertexFormat(format, elementsOf(format), shd);
	STX_FNLN;
	texture=IRenderer::GetRendererInstance()->addImageLibTexture("/ViewportProjectionContent/bluetexture.png", false, IRenderer::GetRendererInstance()->Getlinear());
	texture=IRenderer::GetRendererInstance()->addImageLibTexture("/ViewportProjectionContent/greentexture.png", false, IRenderer::GetRendererInstance()->Getlinear());
	texture=IRenderer::GetRendererInstance()->addImageLibTexture("/ViewportProjectionContent/redtexture.png", false, IRenderer::GetRendererInstance()->Getlinear());
	texture=IRenderer::GetRendererInstance()->addImageLibTexture("/test.bmp", false, IRenderer::GetRendererInstance()->Getlinear());
	STX_FNLN;

	{Image3 img;ubyte* p;char* fn;
	fn="/ViewportProjectionContent/bluetexture.png";	
	img.loadImageLibImage(fn);
	img.decompress();
	p=img.getPixels();
	STX_PRINT("b[0]:%x\n", *(p+0));
	STX_PRINT("b[1]:%x\n", *(p+1));
	STX_PRINT("b[2]:%x\n", *(p+2));}
	{Image3 img;ubyte* p;char* fn;
	fn="/ViewportProjectionContent/greentexture.png";	
	img.loadImageLibImage(fn);
	img.decompress();
	p=img.getPixels();
	STX_PRINT("g[0]:%x\n", *(p+0));
	STX_PRINT("g[1]:%x\n", *(p+1));
	STX_PRINT("g[2]:%x\n", *(p+2));}
	{Image3 img;ubyte* p;char* fn;
	fn="/ViewportProjectionContent/redtexture.png";	
	img.loadImageLibImage(fn);
	img.decompress();
	p=img.getPixels();
	STX_PRINT("r[0]:%x\n", *(p+0));
	STX_PRINT("r[1]:%x\n", *(p+1));
	STX_PRINT("r[2]:%x\n", *(p+2));}
	STX_FNLN;
	
	return 0;
}

void draw()
{
	STX_FNLN;
	IRenderer::GetRendererInstance()->setShader(shd);
	IRenderer::GetRendererInstance()->setVertexFormat(vf);
	IRenderer::GetRendererInstance()->setTexture("Base", texture);
	D3DXFROMWINEVECTOR4 color(0.6f, 0.6f, 0.6f, 1.0f);
	IRenderer::GetRendererInstance()->setShaderConstant4f("color", color);
	D3DXFROMWINEMATRIX I;
	D3DXFROMWINEMatrixIdentity(&I);
	IRenderer::GetRendererInstance()->setShaderConstant4x4f("worldViewProj", I);
	IRenderer::GetRendererInstance()->setDepthState(IRenderer::GetRendererInstance()->GetnoDepthTest());
	#if 1
	//		           x	 y     tx    ty
	float v[16] =  {  1.0f,-1.0f, 1.0f, 1.0f,
			          1.0f, 1.0f, 1.0f, 0.0f,
			         -1.0f,-1.0f, 0.0f, 1.0f,
			         -1.0f, 1.0f, 0.0f, 0.0f };
	unsigned int N=4;
	#else
	float v[16] =  {  1.0f,-1.0f,
			           1.0f, 1.0f,
			          -1.0f,-1.0f,
			          -1.0f, 1.0f };
	unsigned int N=2;
	#endif
#if 1
	IRenderer::GetRendererInstance()->DrawPrimitiveUP(PRIM_TRIANGLE_STRIP, 2, &v[0], &v[0], N*sizeof(float));
#else
	__WORD__ i[6] ={0,1,2,2,1,3};
	IRenderer::GetRendererInstance()->DrawIndexedPrimitiveUP(PRIM_TRIANGLES, 0, 4, 2, &i[0], &i[0], CONSTANT_INDEX2, &v[0], &v[0], N*sizeof(float));
#endif
	STX_FNLN;
}

void render()
{
	STX_FNLN;
	float f=128.0f/256.0f;
	IRenderer::GetRendererInstance()->Clear(true, true, D3DXFROMWINEVECTOR4 (f, f, f, 1.0f));
	STX_FNLN;
	IRenderer::GetRendererInstance()->BeginScene();
	STX_FNLN;
	draw();
	STX_FNLN;
	IRenderer::GetRendererInstance()->EndScene();
	STX_FNLN;
	IRenderer::GetRendererInstance()->Present( );
	STX_FNLN;
}

int ApplicationLogic()
{
	STX_FNLN;
	IRenderer* r=IRenderer::GetRendererInstance("SimpleTexture");
	STX_FNLN;
	IInput*    i=STX_Service::GetInputInstance();
	STX_FNLN;
	init("");
	STX_FNLN;
	while (!i->OnKeyPress (STX_KEY_ESCAPE) && !i->Quit())
	{
	STX_FNLN;
		i->Update();
	STX_FNLN;
		render();
	STX_FNLN;
		//stx_exit(0);
	}
	STX_FNLN;
	return 0;
}

